services/wallet/thirdparty/coingecko/client.go

```diff
@@ -96,13 +101,35 @@ func (c *Client) DoQuery(url string) (*http.Response, error) {
 	return resp, nil
 }
 
+func mapTokensToSymbols(tokens []GeckoToken, tokenMap map[string]GeckoToken) {
+	for _, token := range tokens {
+		if id, ok := coinGeckoMapping[strings.ToUpper(token.Symbol)]; ok {
+			if id != token.ID {
+				continue
+			}
+		}
+		tokenMap[strings.ToUpper(token.Symbol)] = token
+	}
+}
+
 func (c *Client) getTokens() (map[string]GeckoToken, error) {
+	fetching := (&c.isTokenFetchPending).Load()
+
+	if fetching {
+		c.fetchTokensMutex.Lock()
```

Comment:
@IvanBelyakoff I am not sure that I understand your intention here? 
I mean... you have locking here and in the line 127 again? Is that what you wanted?

Reply:
hmm also not liking how the mutex is managed.

Reply:
I don't think you gain much by using the atomic + the mutex, and it makes the logic more complex. In any case, it sounds like premature optimization.
In the end, it seems you're treating both resources (the tokens map and the remote endpoint) as a single entity you want to protect from concurrent access, so simply locking the mutex at the top should do the trick IMO.